core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -59,27 +59,45 @@ private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolea
   // This method should be kept in sync with
   // org.apache.spark.network.shuffle.ExecutorDiskUtils#getFile().
   def getFile(filename: String): File = {
-    // Figure out which local directory it hashes to, and which subdirectory in that
     val hash = Utils.nonNegativeHash(filename)
-    val dirId = hash % localDirs.length
-    val subDirId = (hash / localDirs.length) % subDirsPerLocalDir
-
-    // Create the subdirectory if it doesn't already exist
-    val subDir = subDirs(dirId).synchronized {
-      val old = subDirs(dirId)(subDirId)
-      if (old != null) {
-        old
-      } else {
-        val newDir = new File(localDirs(dirId), "%02x".format(subDirId))
-        if (!newDir.exists() && !newDir.mkdir()) {
-          throw new IOException(s"Failed to create local dir in $newDir.")
+    var count = 0
+    val localDirIndex = scala.collection.mutable.ArrayBuffer(localDirs.indices: _*)
+    var subDir: Option[File] = None
+    while (subDir.isEmpty && count < localDirs.length) {
+      // Figure out which local directory it hashes to, and which subdirectory in that
+      val hashIndex = hash % localDirIndex.size
+      val dirId = localDirIndex(hashIndex)
+      val subDirId = (hash / localDirs.length) % subDirsPerLocalDir
+      // Create the subdirectory if it doesn't already exist
+      subDirs(dirId).synchronized {
+        val old = subDirs(dirId)(subDirId)
+        if (old != null) {
+          subDir = Some(old)
+        } else {
+          val newDir = new File(localDirs(dirId), "%02x".format(subDirId))
+          try {
+            if (!newDir.exists()) {
+              // If parent dir not exist, create it
+              newDir.mkdirs()
+            }
+            subDirs(dirId)(subDirId) = newDir
+            subDir = Some(newDir)
+          } catch {
+            case e: IOException =>
+              logError(s"Failed to create local dir in $newDir.", e)
+              count = count + 1
+              localDirIndex.remove(hashIndex)
```

Comment:
Methinks this change corrupts shuffle writer an reader's dependency

Reply:
> Methinks this change corrupts shuffle writer an reader's dependency

Retry only happened for newSubDir, won't destroy origin dependency. 
If one blockId have corresponding folder. It will return  origin `old`.
For new coming blockId, will retry when mkdir failed.
And if finally return `File` and put into `subDirs`
```
subDirs(dirId)(subDirId) = newDir
```
When this block's request come again, will return `old` since  `subDirs(dirId)(subDirId) != null`

Reply:
the `dirId` seems unstable, I guess for external shuffle probably go wrong

Reply:
> the `dirId` seems unstable, I guess for external shuffle it probably goes wrong

Got your point. If the disk problem was fixed, the `dirId` may not same

Reply:
@yaooqinn Add a method to control this problem, but it doesn't look very elegant. As @srowen mentioned, we can use blacklist to prevent this bad case. But i don't think it can handle all executor disk problem